### PR TITLE
use esword's #41 contribution

### PR DESCRIPTION
**Pull request by nictrix**
A couple things to note:
- would like input from @esword (#41)
  - I would like to discuss if we should keep cli tools and inspec-aws in or should a separate container inherit from this one
- running on the 0.3.0 branch I had to do this to get it to run: `mkdir -p .kitchen/kitchen-terraform/example-ubuntu` - would like to remove that once that issue is fixed
- would like to move the branch in the scripts to master or maybe base it on the current `git status`

Also I would say this is just an experiment at the moment and has no relevance on releasing 0.3.0 version. Though this may be one way to distribute kitchen-terraform. (total image size: 515.7 MB, can get much smaller)

### CODE REVIEW

**Comment by esword** [comment]
Many thanks for producing a PR.  I get torn between having to continue on stuff internally and taking more time to tune it for general purpose use.  

To answer your two questions:
# Tools to include
Yes, the image is currently huge.  If you just want an example docker image using the K-T plugin, you can strip it down a bit.  However, I think it will still be around 400M unless you strip down Ruby.  So if it's going to be that big anyway, why not produce an image with a complete suite of tools so that is actually useful for production testing?  

This may be easier to understand if I explain our primary use case (so far) for the container - We use the image to test reusable TF modules that my team is producing.  The goal is to use it both for rapid iteration as we write the modules and then also invoke it as part of our CircleCI build.  The aws cli is invaluable during the former part of that.  If there is a bug in the module and TF stops, we sometimes need to make direct AWS calls to cleanup resources.  Or we make direct calls to query state as we're debugging.  That's also why there are some additional bash tools in the image.  For example, say you need to tear down a bunch of instances, subnets, etc.  We have a bash script which does this:
```
aws ec2 describe-instances  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.Reservations|.[]|.Instances|.[]|.InstanceId' | xargs -L1 aws ec2 terminate-instances --instance-id
aws ec2 delete-key-pair --key-name kitchen-terraform-example
aws ec2 describe-security-groups  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.SecurityGroups|.[]|.GroupId' | xargs -L1 aws ec2 delete-security-group --group-id
aws ec2 describe-subnets  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.Subnets|.[]|.SubnetId' | xargs -L1 aws ec2 delete-subnet --subnet-id
aws ec2 describe-route-tables  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.RouteTables|.[]|.RouteTableId' | xargs -L1 aws ec2 delete-route-table --route-table-id
aws ec2 describe-internet-gateways  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.InternetGateways|.[]|.InternetGatewayId' | xargs -I igid aws ec2 detach-internet-gateway --internet-gateway-id igid --vpc-id $(aws ec2 describe-vpcs  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.Vpcs|.[]|.VpcId' )
aws ec2 describe-internet-gateways  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.InternetGateways|.[]|.InternetGatewayId' | xargs -L1 aws ec2 delete-internet-gateway --internet-gateway-id
aws ec2 describe-vpcs  --filters "Name=tag:tf-kitchen,Values=true" | jq -r '.Vpcs|.[]|.VpcId' | xargs -L1 aws ec2 delete-vpc --vpc-id
```

(BTW - that is targeted at a slightly tweaked version of your example.  All we did was add the `tf-kitchen` tag to all the resources to make them easier to find.)

So I think the tools are useful.  Though I wouldn't be adverse to splitting into two images if you thought that was best - one that just had the core K-T plugin and one that built on that with all the other tools for interactive debugging. 

# Usage
We have setup a test hierarchy with different directories for each module.  Each dir has minimal scaffolding code to invoke the module under test.  Each dir has a `.kitchen.yml` file and tests to run for that module.  This is likely different than others who may be testing a specific production config and so may have the kitchen files in a separate directory from that being tested.  

The `make` file for the modules project has targets that let us interactively test each module or have circleCI run through all the tests automatically.  The relevant parts look like this.  The module in this case is called `primary-module`:
```
ROOT_DIR := `pwd`
BUILD_DIR := $(ROOT_DIR)/build
TEST_SRC_DIR := $(ROOT_DIR)/test
TEST_OUT_DIR := $(BUILD_DIR)/test
TEST_RUN_DIR := $(TEST_OUT_DIR)/src
SSH_CREDS_DIR := $(TEST_OUT_DIR)/creds/.ssh
AWS_CREDS_DIR := $(TEST_OUT_DIR)/creds/.aws

clean:
	rm -rf $(BUILD_DIR)

build_tests:
	mkdir -p $(TEST_OUT_DIR)
	rsync -av $(TEST_SRC_DIR)/ $(TEST_RUN_DIR)

make_test_ssh_creds:
	$(TEST_RUN_DIR)/scripts/cert_gen.sh $(SSH_CREDS_DIR)

copy_test_aws_creds:
	mkdir -p $(AWS_CREDS_DIR)
	cp $(HOME)/.aws/* $(AWS_CREDS_DIR)
	chmod 0644 $(AWS_CREDS_DIR)/*

run_test_container:
	$(TEST_RUN_DIR)/scripts/run.sh $(ROOT_DIR) "primary-module" "/bin/bash"

#run test container
rtc: run_test_container

clean_test_build: clean build_tests make_test_ssh_creds copy_test_aws_creds

#run new container
rnc: clean_test_build run_test_container

test_primary_module: clean_test_build
	$(TEST_RUN_DIR)/scripts/run.sh $(ROOT_DIR) "primary-module" "./run_suite.sh"

```
So `run_test_container` lets us work in the container interactively, while `test_primary_module` is invoked from our circleCI build.

You can see that all of the test code is first copied into a `build` directory.  That way, all the output files (logs, plans, and state) are not created in the source tree.  It's also easier to clean things up.

`run.sh` looks like this:
```
PROJECT_DIR=$1
TEST_TO_RUN=$2
CMD=$3

set -euo -x pipefail

AWS_DEFAULT_PROFILE=${AWS_DEFAULT_PROFILE:-'default'}
AWS_DEFAULT_REGION=${AWS_DEFAULT_REGION:-'us-east-1'}
# Terraform doesn't respect the AWS_DEFAULT_PROFILE var.  It looks for AWS_PROFILE instead
AWS_PROFILE=$AWS_DEFAULT_PROFILE

TEST_OUT_DIR=$PROJECT_DIR/build/test
CREDS_DIR="$TEST_OUT_DIR/creds"

# These are in the context of the container
PROJECT_MOUNT_PATH="/root/project"
TEST_ROOT_PATH="$PROJECT_MOUNT_PATH/build/test/src"
TEST_PATH="$TEST_ROOT_PATH/$TEST_TO_RUN"

docker run -it --rm  -w "$TEST_PATH" \
-v "$PROJECT_DIR":"$PROJECT_MOUNT_PATH" \
-v "$CREDS_DIR/.ssh":/root/.ssh \
-v "$CREDS_DIR/.aws":/root/.aws \
-e AWS_DEFAULT_PROFILE=$AWS_DEFAULT_PROFILE \
-e AWS_PROFILE=$AWS_PROFILE \
-e AWS_DEFAULT_REGION=$AWS_DEFAULT_REGION \
--name lib_test \
docker.ourrepo.build/ourgroup/tf-tester:latest "$CMD"
```

Walking through in detail:
1. PROJECT_DIR - the root of the project tree - I will likely change this to be the root of the build dir once I do a little more cleanup.
2. CREDS_DIR - A directory containing newly generated ssh keys  (for creating a key pair in AWS) and copies of aws creds
3. TEST_TO_RUN - the subdirectory with the module test code to run.

Clear as mud?

**Comment by esword** [comment]
That was all a long winded way of saying, "Having a run.sh file which mounts the proper directory containing the test code is handy and cleaner than the inline copying that the `run_detailed.sh` does".   Sorry for not being more concise.  That whole set of things evolved from the simple project I setup first to create and test the image.

**Comment by nictrix** [comment]
Thanks for the information! Sounds like a lot of setup for testing. Wondering how and where kitchen-terraform could help in this scenario. I'm sure you're not alone in this testing setup.

I also really liked having the ability to run the detailed example project we have with the container (thinking about using it as part of our CI runs for kitchen-terraform). But if we were to publish a container or provide a way to use kitchen-terraform via a container I'd like to have only whats required in it.

I tried using `-v` for mounting the detailed example, but ran into problems because I'm using Docker Toolbox on macOS.

**Comment by esword** [comment]
Ya, it's a lot of setup for now, I agree.  We're a bit different in that we're not testing one particular setup - we're testing modules that will be used in a variety of setups.  Though I'm hoping to get more teams to start using it for more normal flows.  I present an initial demo tomorrow to try to get folks excited.

I switched to using native Mac docker right at the beginning of this project.  Turned out to be very easy and it's much faster.  You can keep both installed on your machine so you don't lose any existing images.

File: examples/quick_start_with_docker/run_detailed.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+echo "Enter AWS ACCESS KEY"
+read aws_access_key
```
Comment (nictrix) [inline examples/quick_start_with_docker/run_detailed.sh]: 100% agree, this was more of a first pass to see what value it can provide to the project.

**Comment by nictrix** [comment]
Makes sense about the modules, you need to test the features and functions that they provide to the other Terraform projects.

Oh yeah I need to upgrade to the native Docker macOS install, thanks for the reminder!

**Comment by ncs-alane** [comment]
@nictrix do we have a plan here?

**Comment by nictrix** [comment]
@ncs-alane I will create a Dockerfile that can be used to provide Terraform and kitchen-terraform from the container.

**Comment by ncs-alane** [comment]
@nictrix can you rebase this against master? I'd like to delete `ncs-alane-0.3.0`.

**Comment by ncs-alane** [comment]
:point_right: :eye:

**Pull request closed without merging**